go/store/prolly/sort/external.go

```diff
@@ -0,0 +1,458 @@
+type tupleSorter struct {
+	keyCmp    func(val.Tuple, val.Tuple) bool
+	files     [][]keyIterable
+	inProg    *keyMem
+	fileMax   int
+	fileCnt   int
+	batchSize int
+	tmpProv   tempfiles.TempFileProvider
+}
+
+func NewTupleSorter(batchSize, fileMax int, keyCmp func(val.Tuple, val.Tuple) bool, tmpProv tempfiles.TempFileProvider) *tupleSorter {
+	if fileMax%2 == 1 {
+		// round down to even
+		// fileMax/2 will be compact parallelism
+		fileMax -= 1
+	}
+	ret := &tupleSorter{
+		fileMax:   fileMax,
+		batchSize: batchSize,
+		keyCmp:    keyCmp,
+		tmpProv:   tmpProv,
+	}
+	ret.inProg = newKeyMem(batchSize)
+	return ret
+}
+
+func (a *tupleSorter) Flush(ctx context.Context) (iter keyIterable, err error) {
+	// don't flush in-progress, just sort in memory
+	a.inProg.sort(a.keyCmp)
+
+	if len(a.files) == 0 {
+		// don't go to disk if we didn't reach a mem flush
+		return a.inProg, nil
+	}
+
+	var iterables []keyIterable
+	iterables = append(iterables, a.inProg)
+	for _, level := range a.files {
+		for _, file := range level {
+			iterables = append(iterables, file)
+		}
+	}
+
+	newF, err := a.newFile()
+	if err != nil {
+		return nil, err
+	}
+	allKeys := newKeyFile(newF, a.inProg.byteLim)
+	m, _ := newFileMerger(ctx, a.keyCmp, allKeys, iterables...)
+	if err := m.run(ctx); err != nil {
+		return nil, err
+	}
+	return allKeys, nil
+}
+
+func (a *tupleSorter) Insert(ctx context.Context, k val.Tuple) (err error) {
+	if !a.inProg.insert(k) {
+		if err := a.flushMem(ctx); err != nil {
+			return err
+		}
+		a.inProg.insert(k)
+	}
+	return
+}
+
+func (a *tupleSorter) flushMem(ctx context.Context) error {
+	// flush and replace |inProg|
+	if a.inProg.Len() > 0 {
+		newF, err := a.newFile()
+		if err != nil {
+			return err
+		}
+		newFile, err := a.inProg.flush(newF, a.keyCmp)
+		if err != nil {
+			return err
+		}
+		a.inProg = newKeyMem(a.batchSize)
+		if len(a.files) == 0 {
+			a.files = append(a.files, []keyIterable{newFile})
+		} else {
+			a.files[0] = append(a.files[0], newFile)
+		}
+		a.fileCnt++
+	}
+	for level, ok := a.shouldCompact(); ok; level, ok = a.shouldCompact() {
+		if err := a.compact(ctx, level); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *tupleSorter) newFile() (*os.File, error) {
+	f, err := a.tmpProv.NewFile("", "key_file_")
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
+}
+
+func (a *tupleSorter) shouldCompact() (int, bool) {
+	for i, level := range a.files {
+		if len(level) >= a.fileMax {
+			return i, true
+		}
+	}
+	return -1, false
+}
+
+// compact halves the number of files, doubling their size
+func (a *tupleSorter) compact(ctx context.Context, level int) error {
+	fileLevel := a.files[level]
+
+	newF, err := a.newFile()
+	if err != nil {
+		return err
+	}
+	outF := newKeyFile(newF, a.batchSize)
+	m, err := newFileMerger(ctx, a.keyCmp, outF, fileLevel[:a.fileMax]...)
+	if err != nil {
+		return err
+	}
+	if err := m.run(ctx); err != nil {
+		return err
```

Comment:
If we add resource finalization, careful not to leak file newF here. `m.run()` should probably guarantee all iterators it made are closed.